### PR TITLE
Changed Categorical to work with multidim p at the logp level

### LINKED ISSUE #2082: Should Categorical broadcast?
What is the intended behavior when passing a 2D array into Categorical?

```
import pymc3 as pm

with pm.Model():
    x = pm.Bernoulli("x", eye(4)[0], shape=4)
    tr = pm.sample(10)


tr["x"].mean(0)
Out[30]: array([ 1.,  0.,  0.,  0.])



with pm.Model():
    x = pm.Categorical("x", eye(4), shape=4)
    tr = pm.sample(10)

tr["x"].mean(0)
Out[35]: array([ 0.,  0.,  0.,  0.])


In [37]: tr["x"].shape
    ...: 
Out[37]: (10, 4)

```

I was somewhat expecting to see [0, 1, 2, 3], assuming some sort of broadcast.

Also: do others find it alarming that the pymc3 Categorical automatically normalizes the input p vector to sum to 1.0?  To me, having an exception on un-normalized input was an important sanity check in pymc2.  This would be particularly true if 2D inputs are tolerated, in which case row vs. column normalization is always an issue.

### PR DESCRIPTION
**Pull request by lucianopaz**
This should close #2082. However, there is one very important point that we need to discuss:

__When should `Categorical.p` be normalized to sum to 1?__

We must do the normalization for backwards compatibility. But normalizing may hide negative values supplied into `p`. I did some changes to `logp` to normalize `p` there, but still check for the negative values that could have been set in `self.p`, but those changes may break earlier codes that relied on the default normalization. I can't come up with a way to handle the edge cases put forward by @kyleabeauchamp, that contained negative values in `p`, still do the normalization well, and also test for weird values set by step_methods proposals.

### CODE REVIEW

**Review by junpenglao — APPROVED** [review]
LGTM. Nicely done.

File: pymc3/distributions/discrete.py [hunk 1]
```diff
@@ -723,21 +730,34 @@ def random(self, point=None, size=None):
...
+                tt.le(abs(tt.sum(p_, axis=-1) - 1), 1e-5))
+            p = p_
+        else:
+            p = p_ / tt.sum(p_, axis=-1, keepdims=True)
+            sumto1 = True
 
         if p.ndim > 1:
-            a = tt.log(p[tt.arange(p.shape[0]), value_clip])
+            # Fancy trick to index the last axis without touching the other
+            a = tt.log((p.T)[value_clip].T)
```
Comment (junpenglao) [inline pymc3/distributions/discrete.py]: There got to be a better way to do two transpose, right? If it is really necessary, maybe it is easier to do `np.swapaxes` or `np.moveaxes`?

Comment (lucianopaz) [reply pymc3/distributions/discrete.py]: Yup, `np.moveaxes` is the best option. I was not aware that it worked well with theano tensors too.

Comment (junpenglao) [reply pymc3/distributions/discrete.py]: Actually I didnt realized it works with `np.moveaxes` as well. So it works well with theano tensor i take it - would it be better use tt.dimshuffle? http://deeplearning.net/software/theano/library/tensor/basic.html#dimshuffle

**Comment by twiecki** [comment]
Oops, I reverted this commit (https://github.com/pymc-devs/pymc3/commit/de91752fcd750ee40e08d212b8569acb3aa1416e) accidentally. Can you reopen?
